lib/spack/spack/build_systems/cargo.py

```diff
@@ -0,0 +1,247 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+
+import inspect
+import json
+import spack
+
+from llnl.util.filesystem import join_path, copy, mkdirp
+from spack.directives import cargo_manifest, conflicts, depends_on, variant
+from spack.package import PackageBase
+from spack.util.executable import Executable
+from spack.util.rust import target_triple_for_spec, RustQuery
+
+
+class CargoPackage(PackageBase):
+    """
+    Specialized class for installing packages that use the Cargo package
+    manager.
+
+    Cargo is the official build tool and package manager of the Rust
+    programming language. It has its own package repository, which this
+    class will automatically pull from while ensuring checksums are respected.
+
+    This package can install both executable files and C-style static and
+    dynamic libraries from a cargo package. It can not install native
+    Rust-style libraries.
+    """
+
+    phases = ['build', 'install']
+
+    build_system_class = 'CargoPackage'
+
+    variant('build_type',
+            default='release',
+            description='Cargo build type',
+            values=('debug', 'release'))
+
+    variant(
+        'prefer_dynamic',
+        default=True,
+        description='Link Rust standard library dynamically'
+    )
+
+    variant(
+        'lto',
+        default='none',
+        description='Link binaries with link-time optimization',
+        values=('none', 'thin', 'fat')
+    )
+
+    conflicts(
+        'lto=thin',
+        when='+prefer_dynamic',
+        msg="LTO is not possible when linking the Rust standard library "
+            "dynamically. Set '~prefer_dynamic' for this package to enable "
+            "LTO."
+    )
+
+    conflicts(
+        'lto=fat',
+        when='+prefer_dynamic',
+        msg="LTO is not possible when linking the Rust standard library "
+            "dynamically. Set '~prefer_dynamic' for this package to enable "
+            "LTO."
+    )
+
+    depends_on('rust', type='build', when='~prefer_dynamic')
+    depends_on('rust', type=('build', 'link'), when='+prefer_dynamic')
```

Comment:
I'm not actually sure if this works, I've seen problems with changing the deptype based on variants.

Reply:
@adamjstewart you are generally correct but this does work when the constraints are mutually exclusive. An example of something that would not work:

```
depends_on(rust, type=build)
depends_on(rust, type=(build, link), when=+foo)
```

Reply:
I added a comment describing why this is ok.